### PR TITLE
document_processor/split_document_to_partitions

### PR DESCRIPTION
**Pull request by Bhuvanesh-Verma**
This PR adds a document processor `RegexPartitioner`.

### CODE REVIEW

File: src/pie_utils/processor/document/split_into_partitions.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    document: DocumentWithPartition,
+    matcher: Callable[[str], Iterable[Match]],
+    label_group_id: int | None = None,  # = 1,
+    label_whitelist: list[str] | None = None,
+    skip_initial_partition: bool = False,  # = True
+) -> Iterator[Span]:
+    """Spans are created starting with the beginning of matching entries end ending with the start
+    of the next matching one or the end of the document.
+
+    Entries with xml node names that are not in label_whitelist, if it is set, will not be
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/split_into_partitions.py:28]: this is not about xml nodes (you can sue matcher created from any regex)

File: src/pie_utils/processor/document/split_into_partitions.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    considered as match, i.e. their content will be added to the previous matching entry. If
+    label_group_id is set, the content of the respective match group will be taken as label.
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/split_into_partitions.py:30]: this is not the case for the current implementation, we create `Span`s without label. However, since this information may be useful, maybe we switch to `LabeledSpan` for partition? What do you think?

File: src/pie_utils/processor/document/split_into_partitions.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+            document.partition.append(partition)
+            partitions_for_doc.append(partition)
+            partition_texts.append(document.text[partition.start : partition.end])
+            partition_lengths.append(partition.end - partition.start)
+
+        if self.collect_statistics:
+            self.update_statistics("num_partitions", len(document.partition))
+            self.update_statistics("partition_lengths", partition_lengths)
+            self.update_statistics("partition_texts", partition_texts)
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/split_into_partitions.py:121]: Ah ok, that was not clear in the issue description. I meant "This should also collect the distribution of the lengths of [...] the full texts (to compare against)" ;)

**Review by ArneBinder — COMMENTED** [review]
Early feedback. ~Also, can we remove the plural `s` from the file names? (we create only one "partition" that consists of multiple "parts")~ EDIT: as discussed, we stick with the plural.

File: src/pie_utils/processor/document/split_into_partitions.py (hunk 3 above)
Comment (Bhuvanesh-Verma) [reply src/pie_utils/processor/document/split_into_partitions.py:121]: by full texts, do you mean length of full document text ?

File: src/pie_utils/processor/document/split_into_partitions.py (hunk 2 above)
Comment (Bhuvanesh-Verma) [reply src/pie_utils/processor/document/split_into_partitions.py:30]: Yeah as we discussed, labels might be useful in doing filtered evaluation on specific part of the document.

File: src/pie_utils/processor/document/split_into_partitions.py (hunk 3 above)
Comment (ArneBinder) [reply src/pie_utils/processor/document/split_into_partitions.py:121]: yes, to see the effect of the partitioning when comparing with `partition_lengths`

Comment (Bhuvanesh-Verma) [reply src/pie_utils/processor/document/split_into_partitions.py]: fixed in [eeb2b72](https://github.com/ArneBinder/pie-utils/pull/21/commits/eeb2b72429079892d22a699f11d496f42ec3922e)

**Comment by Bhuvanesh-Verma** [comment]
[ed65e59](https://github.com/ArneBinder/pie-utils/pull/21/commits/ed65e59181c21694b93d15c4cd3483353c061f61) adds a label to the initial partition of the document. This is required otherwise initial part do not have any label and is labeled as `None` which results in an exception. 
[7680b52](https://github.com/ArneBinder/pie-utils/pull/21/commits/7680b52f9f5ddc32313d2c35b7327f50d1ccf62e) adds tests for remaining test coverage. However, `test_split_document_to_partitions_with_overlap` is still missing. I tried various patterns to have a overlapped partitions but couldn't work out, any suggestions ?

File: src/pie_utils/processor/document/split_into_partitions.py (hunk 3 above)
Comment (ArneBinder) [reply src/pie_utils/processor/document/split_into_partitions.py]: I think there is an error in the implementation. can you add a check to the test that checks the actual content to `document_lengths` (not just the sum)?

**Comment by ArneBinder** [comment]
> [ed65e59](https://github.com/ArneBinder/pie-utils/pull/21/commits/ed65e59181c21694b93d15c4cd3483353c061f61) adds a label to the initial partition of the document. This is required otherwise initial part do not have any label and is labeled as `None` which results in an exception.

Nice catch. But that should be configurable via a parameter `initial_partition_label: str = "initial part"`.

File: src/pie_utils/processor/document/split_into_partitions.py (hunk 3 above)
Comment (Bhuvanesh-Verma) [reply src/pie_utils/processor/document/split_into_partitions.py]: added this check in [ed39659](https://github.com/ArneBinder/pie-utils/pull/21/commits/ed39659f1b3ba9e0ef2235a6c25474a558f999d9). Is this what you mean by implementation error or something else ?

**Comment by Bhuvanesh-Verma** [comment]
>  EDIT: Does LabeledSpan also accept None as label? If so, use None as default for initial_partition_label.

No `None` is not allowed as label. See [here](https://github.com/ChristophAlt/pytorch-ie/blob/d6bfc925d16da4a94629673fd0cd9a8c155a71de/src/pytorch_ie/annotations.py#L7).

**Comment by ArneBinder** [comment]
> [7680b52](https://github.com/ArneBinder/pie-utils/pull/21/commits/7680b52f9f5ddc32313d2c35b7327f50d1ccf62e) adds tests for remaining test coverage. However, `test_split_document_to_partitions_with_overlap` is still missing. I tried various patterns to have a overlapped partitions but couldn't work out, any suggestions ?

Hmm I also don't see how that can happen. In general, sth like that should be tested in a test for `get_partitions_with_matcher`. Can you add that instead? Then, I think, it will be save to remove the sanity check.

File: tests/processor/document/test_split_into_partition.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    assert len(partitions) == 2
+    partition_lengths = split_document_to_partitions_with_statistics._statistics[
+        "partition_lengths"
+    ]
+    assert partition_lengths == [61, 67, 39, 32, 23]
+    num_partitions = split_document_to_partitions_with_statistics._statistics["num_partitions"]
+    assert num_partitions == [3, 2]
+    document_lengths = split_document_to_partitions_with_statistics._statistics["document_lengths"]
+    assert document_lengths == [len(TEXT1), len(TEXT2)]
+    assert sum(document_lengths) == sum(partition_lengths)
```
Comment (ArneBinder) [inline tests/processor/document/test_split_into_partition.py:79]: That should not be the case if `skip_initial_partition=True`. Can you modify the input text accordingly?

File: tests/processor/document/test_split_into_partition.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+TEXT2 = """<start>Lily is mother of Harry.
+<end>Beth greets Emma.
+"""
+
+
+def test_split_document_to_partitions():
+    split_document_to_partitions = SplitDocumentToPartitions(
+        pattern="(<start>|<middle>|<end>)",
+        label_group_id=0,
+        label_whitelist=["<start>", "<middle>", "<end>"],
```
Comment (ArneBinder) [inline tests/processor/document/test_split_into_partition.py:20]: Can you test for the actual partition labels?

**Review by ArneBinder — CHANGES_REQUESTED** [review]
Some more feedback.

File: src/pie_utils/processor/document/split_into_partitions.py (hunk 3 above)
Comment (ArneBinder) [reply src/pie_utils/processor/document/split_into_partitions.py]: ok, sorry, I misread the code, my bad. But still the check looks better now :)

File: tests/processor/document/test_split_into_partition.py (hunk 5 above)
Comment (Bhuvanesh-Verma) [reply tests/processor/document/test_split_into_partition.py:20]: Do you mean using H1, Title etc as label ? Also, in all examples or just this one ?

Comment (ArneBinder) [reply tests/processor/document/test_split_into_partition.py:20]: no, just check that the `partitions[0].label == "<start>"` etc

Comment (ArneBinder) [reply tests/processor/document/test_split_into_partition.py:20]: and one other thing: can you not include all labels in `label_whitelist`? so that we really test that non-whitelist labels are not used.

Comment (Bhuvanesh-Verma) [reply tests/processor/document/test_split_into_partition.py:27]: fixed in [18bbfe8](https://github.com/ArneBinder/pie-utils/pull/21/commits/18bbfe8640dbbaa8d8108162836663e133dc5d11)

File: tests/processor/document/test_split_into_partition.py (hunk 4 above)
Comment (Bhuvanesh-Verma) [reply tests/processor/document/test_split_into_partition.py]: fixed in [2fff11f](https://github.com/ArneBinder/pie-utils/pull/21/commits/2fff11f7dfcdb6af43c8738f293c0b1c189792d4)

File: src/pie_utils/processor/document/split_into_partition.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations
+
+import json
+import logging
+import re
+from typing import Any, Callable, Iterable, Iterator, Match
+
+from pytorch_ie.annotations import LabeledSpan, Span
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/split_into_partition.py:8]: Is Span still used anywhere?

**Review by ArneBinder — CHANGES_REQUESTED** [review]
I think only the parameter initial_span_label and proper usage test for label_whitelist (see above) is missing (except from docs).

File: tests/processor/document/test_split_into_partition.py (hunk 5 above)
Comment (Bhuvanesh-Verma) [reply tests/processor/document/test_split_into_partition.py]: All cases for `label_whitelist` is tested with [250f47f](https://github.com/ArneBinder/pie-utils/pull/21/commits/250f47f7c9f8b97f4541d39e312f7654a722c2ee) commit. There is a case if we pass empty list as `label_whitelist` which will result in no partitions except for maybe initial_part if that is not skipped. Should we test this case as well ?

File: src/pie_utils/processor/document/split_into_partition.py (hunk 6 above)
Comment (Bhuvanesh-Verma) [reply src/pie_utils/processor/document/split_into_partition.py:8]: Yeah this can be removed now, an extra variable from sanity check prevented it from being removed while optimizing imports.

File: tests/processor/document/test_split_into_partition.py (hunk 5 above)
Comment (ArneBinder) [reply tests/processor/document/test_split_into_partition.py]: good catch, yes, let's add the case where no parts are added. This can also happen, if the matcher does not return any match.

File: src/pie_utils/processor/document/split_into_partition.py (hunk 6 above)
Comment (ArneBinder) [reply src/pie_utils/processor/document/split_into_partition.py:8]: What does this mean?

Comment (Bhuvanesh-Verma) [reply src/pie_utils/processor/document/split_into_partition.py]: I mean while removing code related to sanity check in `split_into_partition.py `, I missed removing a variable of type `Span` that was being used in that code block. Its fixed in [ae085da](https://github.com/ArneBinder/pie-utils/pull/21/commits/ae085da985e50e43213cb7c97299bd4dc52e9c07).

File: tests/processor/document/test_split_into_partition.py (hunk 5 above)
Comment (Bhuvanesh-Verma) [reply tests/processor/document/test_split_into_partition.py]: Test cases added in [b461ce0](https://github.com/ArneBinder/pie-utils/pull/21/commits/b461ce0d6b85d1a4308ba630ff361bfc5b6d290d)

File: src/pie_utils/processor/document/split_into_partition.py (hunk 6 above)
Comment (ArneBinder) [reply src/pie_utils/processor/document/split_into_partition.py]: ah ok, thanks!

File: tests/processor/document/test_split_into_partition.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        label_whitelist=[],
+    )
+    document = DocumentWithPartition(text=TEXT1)
+    new_document = split_document_to_partitions(document)
+
+    partitions = new_document.partition
+    assert len(partitions) == 1
+    partition = partitions[0]
+    assert partition.label == "initial part"
+    assert new_document.text[partition.start : partition.end] == TEXT1
```
Comment (ArneBinder) [inline tests/processor/document/test_split_into_partition.py:171]: Ok, this needs to be well document. The behavior looked a bit strange at the beginning, when compared to the other outputs for no matches, but I guess, it is intended. Or do you think it makes sense to change that?

File: tests/processor/document/test_split_into_partition.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+
+def test_split_document_to_partitions_with_no_parts_added():
+    split_document_to_partitions = SplitDocumentToPartitions(
+        pattern="(<start>|<middle>|<end>)",
+        label_group_id=0,
+        label_whitelist=[],
+        skip_initial_partition=True,
+    )
+    document = DocumentWithPartition(text=TEXT1)
```
[the remark this answers is not in the document]
Comment (ArneBinder) [reply tests/processor/document/test_split_into_partition.py:153]: maybe let's do that for all tests

File: tests/processor/document/test_split_into_partition.py (hunk 7 above)
Comment (Bhuvanesh-Verma) [reply tests/processor/document/test_split_into_partition.py:171]: Yeah so what I thought that it would be useful if we want whole document as a partition then we can do it this way with a custom label to partition. So I think it is fine as it is at the moment. I'll add necessary documentation.

Comment (ArneBinder) [reply tests/processor/document/test_split_into_partition.py:171]: sounds good!

File: tests/processor/document/test_split_into_partition.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Bhuvanesh-Verma) [reply tests/processor/document/test_split_into_partition.py:153]: maybe I can define TEXT as variable in each test, since it is used in other places as well. If we write whole text everywhere then it'll expand the script.

[the remark this answers is not in the document]
Comment (ArneBinder) [reply tests/processor/document/test_split_into_partition.py:153]: yeah, I meant that

File: src/pie_utils/processor/document/split_into_partition.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            previous_start = match.start()
+            previous_label = label
+
+    if previous_start is not None:
+        end = len(document.text)
+        span = LabeledSpan(start=previous_start, end=end, label=previous_label)
+        yield span
+
+
+class SplitDocumentToPartitions(WithStatistics):
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/split_into_partition.py:63]: Let's rename that to `RegexPartitioner` and the files (test_)regex_partitioner.py.

File: tests/processor/document/test_split_into_partition.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Bhuvanesh-Verma) [reply tests/processor/document/test_split_into_partition.py]: fixed in [d7d4e3c](https://github.com/ArneBinder/pie-utils/pull/21/commits/d7d4e3c0dbebdbd3bd104877d44c2b19cb9f53f3)

File: src/pie_utils/processor/document/split_into_partition.py (hunk 9 above)
Comment (Bhuvanesh-Verma) [reply src/pie_utils/processor/document/split_into_partition.py]: renamed in [906106e](https://github.com/ArneBinder/pie-utils/pull/21/commits/906106e0237869813aa0d3fe3070b05a231b804a)

File: tests/processor/document/test_split_into_partition.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ArneBinder) [reply tests/processor/document/test_split_into_partition.py]: thanks!

File: src/pie_utils/processor/document/regex_partitioner.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    def reset_statistics(self):
+        self._statistics: dict[str, Any] = {
+            "partition_lengths": [],
+            "num_partitions": [],
+            "document_lengths": [],
+        }
+
+    def show_statistics(self, description: str | None = None):
+        description = description or "Statistics"
+        logger.info(f"{description}: \n{json.dumps(self._statistics, indent=2)}")
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/regex_partitioner.py:90]:
```suggestion
        statistics_show = {
          key: {
            "min": min(values),
            "max": max(values),
            "mean": statistics.mean(values),
            "stddev": statistics.pstdev(values),
            } for k, v in self._statistics.items()}

        logger.info(f"{description}: \n{json.dumps(statistics_show, indent=2)}")
```

**Review by ArneBinder — CHANGES_REQUESTED** [review]
Totally forget, that otherwise the statistics would not make much sense (hard to read long lists of ints))

File: src/pie_utils/processor/document/split_into_partitions.py (hunk 1 above)
Comment (Bhuvanesh-Verma) [reply src/pie_utils/processor/document/split_into_partitions.py]: fixed in [a9eb530](https://github.com/ArneBinder/pie-utils/pull/21/commits/a9eb5309bb21ae4a6ace9e3dbcadab2869a1c537)

File: src/pie_utils/processor/document/regex_partitioner.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    label_group_id: int | None = None,  # = 1,
+    label_whitelist: list[str] | None = None,
+    skip_initial_partition: bool = False,  # = True
+    initial_partition_label: str = "initial part",
+) -> Iterator[LabeledSpan]:
+    """This method yields LabeledSpans as partitions of the given document. matcher is used to
+    search for a pattern in the document. If pattern is found then it returns Match objects that
+    contains matched groups. The required match group is selected using label_group_id to create a
+    partition label. The span of the partition ranges between two matched pattern. It should be
+    noted that none of the partitions overlap.
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/regex_partitioner.py:30]: * There are articles missing. 
* Also, logic for setting the label is optional (do we have a test for `label_group_id=None`?) and, hence, should be explained at the end (especially before the main functionality). 
* It is also not correct, that "The span of the partition ranges _between_ two matched pattern".
* regarding "It should be noted that none of the partitions overlap.": This sentence does not make sense to me, as said previously, I dont think we have multiple partitions (but multiple parts). But maybe I am really wrong here. Or do you have ideas for better naming? Renaming partition to partitioning?

File: src/pie_utils/processor/document/regex_partitioner.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    3. There will be no partitions if
+        a. matcher could not find pattern in document text
+        b. label_whitelist is empty list and skip_initial_partition is True
+        c. label_group_id and label_whitelist is None and skip_initial_partition is True
+        d. label_group_id is None and skip_initial_partition is False
+
+    :param document: A Document that is to be partitioned
+    :param matcher: A method that is used to match a pattern in the document text and return an iterator yielding the
+                    Match objects.
+    :param label_whitelist: A list of labels which are allowed to form a partition. (the default value is None)
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/regex_partitioner.py:46]: ...an optional list...

Comment (ArneBinder) [reply src/pie_utils/processor/document/regex_partitioner.py:46]: say that it only makes sense when label_group_id is not None

File: src/pie_utils/processor/document/regex_partitioner.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    :param label_group_id: An integer value to select suitable match group from Match object. (the default value is None)
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/regex_partitioner.py:47]: to do what? what happens, if None (default)?

Comment (ArneBinder) [reply src/pie_utils/processor/document/regex_partitioner.py:47]: should be before `label_whitelist`

File: src/pie_utils/processor/document/regex_partitioner.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+    :param skip_initial_partition: A boolean value that prevents initial partition to be saved in the document. (the
+                                default value is False)
+    :param initial_partition_label: A string value used as a partition label for initial partition. This is only used when
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/regex_partitioner.py:50]: missing default. general for defaults: you can just have "(default: value)" after it is first mentioned, e.g.: "A string value (default: "initial part") used as a  partition label ..."

File: src/pie_utils/processor/document/regex_partitioner.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+            previous_label = label
+
+    if previous_start is not None and previous_label is not None:
+        end = len(document.text)
+        span = LabeledSpan(start=previous_start, end=end, label=previous_label)
+        yield span
+
+
+class RegexPartitioner(WithStatistics):
+    """RegexPartitioner partitions a document into multiple parts using a regular expression.
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/regex_partitioner.py:80]: important information to add: the matched string is included in the parts (at the beginning)!

Comment (ArneBinder) [reply src/pie_utils/processor/document/regex_partitioner.py:80]: also, at least refer to the method `get_partitions_with_matcher` for further information.

**Review by ArneBinder — CHANGES_REQUESTED** [review]
Early feedback regarding the docs (no complete review)

**Comment by Bhuvanesh-Verma** [comment]
All above requested changes are done in [1a80981](https://github.com/ArneBinder/pie-utils/pull/21/commits/1a809817ad0cf18b00a0c2026916ab6efc6cd8fa).
Tests related to `label_group_id` and `no match found` are parameterized in [011cca5](https://github.com/ArneBinder/pie-utils/pull/21/commits/011cca586e807f0401f9467961ce0300433f5db0) and respective documentation is added in [78ae731](https://github.com/ArneBinder/pie-utils/pull/21/commits/78ae7317ad86e8745a27daefeae3649dd97ed86b).

File: src/pie_utils/processor/document/regex_partitioner.py [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+    label_whitelist: list[str] | None = None,
+    skip_initial_partition: bool = False,  # = True
+    default_partition_label: str = "partition",
+    initial_partition_label: str | None = None,
+) -> Iterator[LabeledSpan]:
+    """This method yields LabeledSpans as partitions of the given document. matcher is used to
+    search for a pattern in the document. If the pattern is found, it returns a Match object that
+    contains matched groups. A partition is then created using a span in the matched groups. The
+    span of a partition starts from the first match (inclusive) and ends at the second match
+    (exclusive). A partition is labeled either using the default_partition_label or using the list
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/regex_partitioner.py:31]:
```suggestion
    span of a partition starts from the first match (inclusive) and ends at the next match
    (exclusive) or at the end of the document. A partition is labeled either using the default_partition_label or using the list
```

File: src/pie_utils/processor/document/regex_partitioner.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    of labels available in label_whitelist. It should be noted that none of the partitions overlap.
+
+    :param document: A Document that is to be partitioned
+    :param matcher: A method that is used to find a pattern in the document text and return an iterator yielding the
+                    Match objects.
```
Comment (ArneBinder) [inline src/pie_utils/processor/document/regex_partitioner.py:36]:
```suggestion
    :param matcher: A method that is used to find a pattern in the document text and return an iterator yielding the
                    Match objects, e.g. re.compile(PATTERN).finditer
```

**Review by ArneBinder — CHANGES_REQUESTED** [review]
just minors

**Review by ArneBinder — APPROVED** [review]
